File: kangaroo_app_sdk_android/android/src/main/kotlin/com/kangaroo/flutterplugin/android/features/redeemAmazonCatalogueItem/RedeemAmazonCatalogueItemsHandler.kt
@file:Suppress("INLINE_FROM_HIGHER_PLATFORM")
@file:OptIn(ExperimentalJsExport::class)
package com.kangaroo.flutterplugin.android.features.redeemAmazonCatalogueItem

import com.kangaroo.flutterplugin.android.base.PluginChannelHandler
import com.kangaroo.flutterplugin.android.base.pushSerializedResultToEventSink
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import kotlin.js.ExperimentalJsExport
import features.redeemAmazonCatalogueItem.RedeemAmazonCatalogueItemsApi
import features.redeemAmazonCatalogueItem.models.RedeemResponseModel

import features.redeemAmazonCatalogueItem.serializeRedeemAmazonCatalogueItemsState
import kangaroorewards.appsdk.core.domain.SerializedResult
import kangaroorewards.appsdk.core.domain.toJsonResult
import kotlinx.serialization.decodeFromString
import kotlinx.serialization.json.Json
import features.redeemAmazonCatalogueItem.models.RedeemAmazonItemRequest

class RedeemAmazonCatalogueItemsHandler : EventChannel.StreamHandler, PluginChannelHandler {
    var sink: EventChannel.EventSink? = null

    override val methodChannel: String
        get() = "customer_sdk/methods/redeem_amazon_catalogue_item"

    override val eventChannel: String
        get() = "customer_sdk/events/redeem_amazon_catalogue_item"

    override suspend fun onMethodCall(call: MethodCall): String? {
        return redeemAmazonCatalogueItem(call)
    }

    override fun getStreamHandler(): EventChannel.StreamHandler {
        return this
    }

    companion object {
        suspend fun redeemAmazonCatalogueItem(call: MethodCall): String? {
            val result = RedeemAmazonCatalogueItemsApi().redeemAmazonCatalogueItem(
                overrideHeaders = call.argument<Map<String, String>>("overrideHeaders") as Map<String, String>?,
                redeemAmazonItemRequest = Json.decodeFromString(call.argument<String>("redeemAmazonItemRequest") as String)
            ).toJsonResult<RedeemResponseModel>()

            return when (result) {
                is SerializedResult.Success -> result.data
                is SerializedResult.UnauthorizedError -> result.error
                is SerializedResult.UnknownError -> result.error
                else -> null
            }
        }
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        sink = events
        RedeemAmazonCatalogueItemsApi().redeemAmazonCatalogueItemsState.serializeRedeemAmazonCatalogueItemsState().observe { result ->
            pushSerializedResultToEventSink<RedeemAmazonCatalogueItemsApi>(sink, result)
        }
    }

    override fun onCancel(arguments: Any?) {
        sink = null
    }
}

File: kangaroo_app_sdk_ios/ios/Classes/features/redeemAmazonCatalogueItem/RedeemAmazonCatalogueItemsHandler.swift
import Foundation
import Flutter
import KangarooAppSdkCustomer

class RedeemAmazonCatalogueItemsHandler: NSObject, FlutterStreamHandler, PluginChannelHandler {
    var sink: FlutterEventSink?

    var methodChannel: String = "customer_sdk/methods/redeem_amazon_catalogue_item"

    var eventChannel: String = "customer_sdk/events/redeem_amazon_catalogue_item"

    func onMethodCall(call: FlutterMethodCall) async -> Any? {
        return await RedeemAmazonCatalogueItemsHandler.redeemAmazonCatalogueItem(call: call)
    }

    func getStreamHandler() -> (FlutterStreamHandler & NSObjectProtocol)? {
        return self
    }


    static func redeemAmazonCatalogueItem(call: FlutterMethodCall) async -> String? {


        
    let args = call.arguments
        do {

        var overrideHeaders: [String: String]?

        if let myArgs = args as? [String: Any] {
            overrideHeaders = myArgs["overrideHeaders"] as? [String: String]
        }
        else {
            overrideHeaders = [:]
        }

        let result = try await RedeemAmazonCatalogueItemsApi().redeemAmazonCatalogueItem(overrideHeaders: overrideHeaders, methods: call.arguments as! [String : Any]).serializeRedeemAmazonCatalogueItemsApiResult()

            switch result {
                case let result as SerializedResultSuccess:
                    return result.data
                case let result as SerializedResultUnauthorizedError:
                    return result.error
                case let result as SerializedResultUnknownError:
                    return result.error
                default:
                    return nil
            }

        }
        catch {
            return nil
        }


        


        return nil
    }

    func onListen(withArguments arguments: Any?, eventSink events: @escaping
        FlutterEventSink) -> FlutterError? {
        sink = events
        RedeemAmazonCatalogueItemsApi().redeemAmazonCatalogueItemsState
            .serializeRedeemAmazonCatalogueItemsState().observe { result in
            switch result {
            case let result as SerializedResultIdle:
                self.sink?(result.state)
            case let result as SerializedResultLoading:
                self.sink?(result.state)
            case let result as SerializedResultSuccess:
                self.sink?(result.data)
            case let result as SerializedResultEmptyResponse:
                self.sink?(result.body)
            case let result as SerializedResultUnauthorizedError:
                self.sink?(result.error)
            case let result as SerializedResultUnknownError:
                self.sink?(result.error)
            default:
                break
            }
        }
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        sink = nil
        return nil
    }
}
